src/pollers/retry.rs

```diff
@@ -0,0 +1,334 @@
+use std::future::Future;
+use std::{fmt::Debug, time::Duration};
+
+use backoff::backoff::Backoff;
+use backoff::{ExponentialBackoff, SystemClock};
+use futures_retry::{ErrorHandler, FutureRetry, RetryPolicy};
+use tonic::Code;
+
+use crate::pollers::default;
+use crate::pollers::gateway::ServerGatewayApis;
+use crate::{
+    pollers::Result,
+    protos::{
+        coresdk::common::Payload,
+        coresdk::workflow_commands::QueryResult,
+        temporal::api::{
+            common::v1::Payloads, enums::v1::WorkflowTaskFailedCause, failure::v1::Failure,
+            query::v1::WorkflowQuery, workflowservice::v1::*,
+        },
+    },
+    protosext::WorkflowTaskCompletion,
+    task_token::TaskToken,
+};
+use std::time::Instant;
+
+#[derive(Debug)]
+pub struct RetryGateway<SG> {
+    gateway: SG,
+}
+
+#[derive(Debug)]
+pub struct TonicErrorHandler {
+    backoff: ExponentialBackoff,
+    max_attempts: usize,
+}
+
+impl TonicErrorHandler {
+    pub fn new(backoff: ExponentialBackoff, max_attempts: usize) -> Self {
+        TonicErrorHandler {
+            backoff,
+            max_attempts,
+        }
+    }
+}
+
+impl ErrorHandler<tonic::Status> for TonicErrorHandler {
+    type OutError = tonic::Status;
+
+    fn handle(&mut self, current_attempt: usize, e: tonic::Status) -> RetryPolicy<tonic::Status> {
+        if current_attempt >= self.max_attempts {
+            return RetryPolicy::ForwardError(e);
+        }
+        match e.code() {
+            Code::Cancelled
+            | Code::DataLoss
+            | Code::DeadlineExceeded
+            | Code::Internal
+            | Code::Unknown
+            | Code::ResourceExhausted
+            | Code::Aborted
+            | Code::OutOfRange
+            | Code::Unimplemented
+            | Code::Unavailable
+            | Code::Unauthenticated => {
+                match self.backoff.next_backoff() {
+                    None => RetryPolicy::ForwardError(e), // None is returned when we've ran out of time.
+                    Some(backoff) => RetryPolicy::WaitRetry(backoff),
+                }
+            }
+            _ => RetryPolicy::ForwardError(e),
+        }
+    }
+}
+
+impl<SG> RetryGateway<SG> {
+    pub fn new(gateway: SG) -> Self {
+        Self { gateway }
+    }
+}
+
+#[async_trait::async_trait]
+impl<SG: ServerGatewayApis + Send + Sync + 'static> ServerGatewayApis for RetryGateway<SG> {
+    async fn start_workflow(
```

Comment:
This whole thing should really just be done with a macro, but we can do that in another PR like we talked about

Reply:
Yes, let's do it as a follow up, would be a good exercise for me to learn macros.